### ISSUE TITLE
C# Debugging

### ISSUE THREAD

**Issue by Nightfall333** [dotnet/vscode-csharp#4599]
When i try to run my code, it says that I need an extension. It then takes me to a list of extensions. I tried them all and none of them work.

**Comment by superman-lopez** [comment]
@Nightfall333 You have installed the "C# for Visual Studio Code" extension, and it's enabled?

**Comment by WardenGnaw** [comment]
The "Error "You don't have an extension of ..." with debugger extension" Dialog comes from VSCode. 

Potentially a duplicate of https://github.com/microsoft/vscode/issues/123881

**Comment by WardenGnaw** [comment]
Actually, still hitting this on 1.57.

<details>
<summary>VS Code Version</summary>

```
Version: 1.57.0 (user setup)
Commit: b4c1bd0a9b03c749ea011b06c6d2676c8091a70c
Date: 2021-06-09T17:18:42.354Z
Electron: 12.0.9
Chrome: 89.0.4389.128
Node.js: 14.16.0
V8: 8.9.255.25-electron.0
OS: Windows_NT x64 10.0.19043
```
</details>

@isidorn @weinand Is there something we are missing in our `package.json`? 

**Repro steps:**
1. Install .NET 5
2. Run `dotnet new console`
3. Open VS Code in this folder
4. F5 the Program.cs file

**Expected:** 
See configuration options

**Actual:**

Error:
![image](https://user-images.githubusercontent.com/3953714/121737078-40306b00-caad-11eb-844c-4869c0c42654.png)
Installed Extensions
![image](https://user-images.githubusercontent.com/3953714/121737088-458db580-caad-11eb-8fd4-68957ccd6792.png)
Extension|Author (truncated)|Version
---|---|---
csharp|ms-|1.23.12
remote-wsl|ms-|0.56.5
cpptools|ms-|1.4.1
powershell|ms-|2021.5.1

**Comment by isidorn** [comment]
@WardenGnaw sorry for the slow response I was on vacation the previous week.
Your extension needs to say that it is interested in the `C#` language in your package.json
Here's an example how mock debug is doing that https://github.com/microsoft/vscode-mock-debug/blob/main/package.json#L135

Can you please try that and let me know how it goes?
Thanks!

**Comment by WardenGnaw** [comment]
> sorry for the slow response I was on vacation the previous week.

No worries! I hope you had a wonderful vacation!

------------------

Was there a change recently that required this? I thought the debugger language was covered by the [`breakpoints`](https://github.com/OmniSharp/omnisharp-vscode/blob/77dee006876d1d5832711e8a981ab928cedb351d/package.json#L1032-L1045) section of the `package.json`.

> Can you please try that and let me know how it goes?

That resolved the prompt, thanks for the suggestion for a quick fix 😃

**Comment by isidorn** [comment]
The `breakpoints` section only controls if breakpoint can be set in a file or not. It is not even part of the `debuggers` contribution because we wanted to keep this separate.

We always had a `debuggers.languages` contribution and we treated it as a way for extension to notify VS Code that they are interested in a language. VS Code would use this knowledge when user tries to debug without a launch.json. For example if the user has a `python` file and presses F5 we would pick the Python debugger to start debugging and figure out what to do even though there is no `launch.json`.

Last milestone there was a change which broke C# - sorry about that! C# seems to have been using an approach that an F5 on a C# file would have to generate launch configs, even though C# never mentioned that it is interested in C# files.
This was working until last milestone, but we made our requirements stronger that in order to support debugging with no launch.json the debug extension now has to specify it supports a language.
This has improved the scenario where a user has no C# extension, the user has a .cs file and presses F5. Now we will prompt the user properly to install the C# extension. Which I think is a very nice flow for new users.

I hope this clarifies why we changed this.

**Comment by weinand** [comment]
@isidorn If a user has the C# extension installed, it does not make a lot of sense to show this dialog:

![121737078-40306b00-caad-11eb-844c-4869c0c42654](https://user-images.githubusercontent.com/1898161/122991835-c8641980-d3a5-11eb-8131-7f83c0f04833.png)

Yes, I understand that the missing "debuggers.languages" contribution prevents us from knowing whether one of the installed debugging extensions will work for C# files.

How can we fix this?
I see these options:

- since our change will break all debugging extensions that lack the "debuggers.languages" contribution, we will have to make sure that all debugger extensions add a proper "debuggers.languages" contribution before we can release the new feature.
- if we don't want to wait until all debuggers have been updated, we could:
  - use the `breakpoints` contribution in a debugger's package.json as a strong indication whether the debugger can handle the file type.
  - show a debugger Quickpick to let the user select one of his debugger extensions, and only if there is none let the user search on the Marketplace.







-

**Comment by weinand** [comment]
I've created https://github.com/microsoft/vscode/issues/126962 for discussing the general issue behind this.

**Comment by JoeRobich** [comment]
Closing as resolved.
